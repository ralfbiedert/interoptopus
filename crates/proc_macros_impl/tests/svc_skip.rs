use quote::quote;
use syn::{ItemImpl, parse_quote};

mod util;

#[test]
fn skip_impl() {
    let item: ItemImpl = parse_quote! {
        #[ffi_service]
        impl Service {
            #[ffi::skip]
            fn new() -> ffi::Result<Self, Error> { }
        }
    };

    insta::assert_snapshot!(expand_svc!(item));
}
