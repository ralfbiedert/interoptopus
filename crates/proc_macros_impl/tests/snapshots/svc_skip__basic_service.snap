---
source: crates/proc_macros_impl/tests/svc_skip.rs
assertion_line: 15
expression: expand_svc!(item)
---
const _: () = {
    const fn _assert_type_info<T: ::interoptopus::lang::types::TypeInfo>() {}
    _assert_type_info::<Service>();
    ::interoptopus::lang::types::assert_service_ctor_safe::<
        ffi::Result<Service, Error>,
    >();
    ::interoptopus::lang::types::assert_service_safe::<Service>();
};
impl Service {
    fn new() -> ffi::Result<Self, Error> {}
}
#[::interoptopus::ffi_function]
unsafe fn service_new(
    instance: *mut *const Service,
) -> <::interoptopus::ffi::Result<
    (),
    Error,
> as ::interoptopus::pattern::result::ResultAs>::AsT<*const Service> {
    unsafe {
        let result = Service::new();
        match result {
            ::interoptopus::ffi::Ok(service_instance) => {
                let boxed = ::std::boxed::Box::new(service_instance);
                *instance = ::std::boxed::Box::into_raw(boxed);
                ::interoptopus::ffi::Ok(::std::ptr::null())
            }
            ::interoptopus::ffi::Err(err) => ::interoptopus::ffi::Err(err),
            ::interoptopus::ffi::Result::Panic => ::interoptopus::ffi::Result::Panic,
            ::interoptopus::ffi::Result::Null => ::interoptopus::ffi::Result::Null,
        }
    }
}
#[::interoptopus::ffi_function]
fn service_destroy(instance: *const Service) {
    if !instance.is_null() {
        unsafe {
            let _ = ::std::boxed::Box::from_raw(instance as *mut Service);
        }
    }
}
impl ::interoptopus::lang::service::ServiceInfo for Service {
    fn id() -> ::interoptopus::inventory::ServiceId {
        ::interoptopus::inventory::ServiceId::from_id(::interoptopus::id!(Service))
    }
    fn service() -> ::interoptopus::lang::service::Service {
        ::interoptopus::lang::service::Service::new(
            <Service as ::interoptopus::lang::types::TypeInfo>::id(),
            vec![< service_new as ::interoptopus::lang::function::FunctionInfo > ::id()],
            <service_destroy as ::interoptopus::lang::function::FunctionInfo>::id(),
            vec![],
        )
    }
    fn register(inventory: &mut ::interoptopus::inventory::Inventory) {
        <Service as ::interoptopus::lang::types::TypeInfo>::register(inventory);
        <service_new as ::interoptopus::lang::function::FunctionInfo>::register(
            inventory,
        );
        <service_destroy as ::interoptopus::lang::function::FunctionInfo>::register(
            inventory,
        );
        inventory.register_service(Self::id(), Self::service());
    }
}
